data_model/src/events/trigger_executed.rs

```diff
@@ -0,0 +1,255 @@
+//! Event that notifies some trigger was somehow executed.
+
+#[cfg(not(feature = "std"))]
+use alloc::{format, string::String, vec::Vec};
+
+use iroha_ffi::FfiType;
+use iroha_macro::FromVariant;
+use iroha_schema::prelude::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+use super::Filter;
+use crate::prelude::{Metadata, TriggerId};
+
+/// [`Event`] filter.
+#[derive(
+    Default,
+    Debug,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Clone,
+    Decode,
+    Encode,
+    IntoSchema,
+    Hash,
+    Serialize,
+    Deserialize,
+    FfiType,
+)]
+pub struct EventFilter {
+    /// If `Some`, filters by the [`TriggerId`]. Otherwise accepts any [`TriggerId`].
+    pub id: Option<TriggerId>,
+    /// If `Some`, filters by the [`StatusKind`]. Otherwise accepts any [`StatusKind`].
+    pub status_kind: Option<StatusKind>,
+}
+
+impl EventFilter {
+    /// Construct [`EventFilter`].
+    #[must_use]
+    #[inline]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Filter by [`TriggerId`].
+    #[must_use]
+    #[inline]
+    pub fn id(mut self, id: TriggerId) -> Self {
+        self.id = Some(id);
+        self
+    }
+
+    /// Filter by [`StatusKind`].
+    #[must_use]
+    #[inline]
+    pub const fn status_kind(mut self, status_kind: StatusKind) -> Self {
+        self.status_kind = Some(status_kind);
+        self
+    }
+
+    #[inline]
+    fn field_matches<T: Eq>(filter: Option<&T>, event: &T) -> bool {
+        filter.map_or(true, |field| field == event)
+    }
+}
+
+impl Filter for EventFilter {
+    type Event = Event;
+
+    /// Check if `self` accepts the `event`.
+    #[inline]
+    fn matches(&self, event: &Event) -> bool {
+        [
+            Self::field_matches(self.id.as_ref(), &event.id),
+            Self::field_matches(self.status_kind.as_ref(), &event.status.kind()),
+        ]
+        .into_iter()
+        .all(core::convert::identity)
+    }
+}
+
+/// Notification that some trigger was somehow executed.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Decode, Encode, Deserialize, Serialize, IntoSchema)]
+pub struct Event {
+    /// Which trigger was executed.
+    pub id: TriggerId,
+    /// How the trigger execution is going on, or resulted in.
+    pub status: Status,
+}
+
+impl Event {
+    /// Construct [`Event`].
+    pub const fn new(id: TriggerId, status: Status) -> Self {
+        Event { id, status }
+    }
+}
+
+/// How the trigger execution is going on, or resulted in.
+#[derive(
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    Hash,
+    Decode,
+    Encode,
+    Serialize,
+    Deserialize,
+    FromVariant,
+    IntoSchema,
+)]
+pub enum Status {
+    /// The trigger execution was succeeded.
+    Succeeded,
+    /// The trigger execution was failed due to [`FailReason`].
+    Failed(FailReason),
+}
+
+/// Details of why the trigger execution failed.
+pub type FailReason = Metadata;
```

Comment:
Defined as the most general-purpose data structure

Reply:
For now, yes. We need to create an Epic to actually produce appropriate error messages.